### PR TITLE
bpo-35803: Document and test dir=PathLike for tempfile

### PR DESCRIPTION
**Pull request by asottile**
https://bugs.python.org/issue35803

### CODE REVIEW

File: Doc/library/tempfile.rst [hunk 1]
```diff
@@ -183,6 +183,10 @@ The module defines the following user-callable items:
       *suffix* and *prefix* now accept and default to ``None`` to cause
       an appropriate default value to be used.
 
+   .. versionchanged:: 3.6
+      The *dir* parameter now accepts objects implementing the
```
Comment (tirkarthi) [inline Doc/library/tempfile.rst]: Perhaps this and below could be simplified with :term:`path-like object`? I haven't checked the markup an example https://docs.python.org/3.8/library/os.html#os.chdir . I see this version on newer docs and hence my personal preference.

The *dir* parameter now accepts :term:`path-like object`

Comment (asottile) [reply Doc/library/tempfile.rst]: done, thanks!

File: Lib/test/test_tempfile.py [hunk 2]
```diff
@@ -84,8 +88,13 @@ def nameCheck(self, name, dir, pre, suf):
         nsuf  = nbase[len(nbase)-len(suf):]
 
         if dir is not None:
-            self.assertIs(type(name), str if type(dir) is str else bytes,
-                          "unexpected return type")
+            self.assertIs(
+                type(name),
+                str
+                if type(dir) is str or isinstance(dir, os.PathLike) else
```
Comment (tirkarthi) [inline Lib/test/test_tempfile.py:94]: Is there a difference between current code and `isinstance(dir, (str, os.PathLike))`?

File: Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Document and test that ``tempfile`` functions may take a ``dir`` which
+is a :term:`path-like object`.  Patch by Anthony Sottile
```
Comment (tirkarthi) [inline Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst]: Minor nit

```suggestion
is a :term:`path-like object`.  Patch by Anthony Sottile.
```

Comment (tirkarthi) [reply Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst]: Also can state the two functions directly here instead of tempfile functions?

Comment (asottile) [reply Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst]: it's not just two functions, it's every function in the `tempfile` module

File: Lib/test/test_tempfile.py (hunk 2 above)
Comment (asottile) [reply Lib/test/test_tempfile.py:94]: ever so slightly.  The current code is "type is exactly `str`", my patch is "type is exactly `str` or it implements `PathLike`", your suggestion is "`str` or any subclass of `str` or implements `PathLike`"

File: Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst (hunk 3 above)
Comment (asottile) [reply Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst]: ah I see what you're saying here.

I only added the `versionchanged` comment to the two functions following the convention for the versionchanged directly above it.  The functions in the `tempfile` module all mention "The dir, prefix and suffix parameters have the same meaning and defaults as with mkstemp()." and don't document those parameters themselves

**Review by JelleZijlstra — APPROVED** [review]

File: Lib/test/test_tempfile.py (hunk 2 above)
Comment (ammaraskar) [inline Lib/test/test_tempfile.py:94]: Maybe break this up to it's own line since its getting a bid crowded as a ternary

File: Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Document and test that ``tempfile`` functions may take a ``dir`` which
```
Comment (ammaraskar) [inline Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst]: Maybe rephrase this to 

```suggestion
Document and test that ``tempfile`` functions may accept a
:term:`path-like object` for the ``dir`` argument.  Patch by Anthony Sottile.
```

right now it sounds like the `dir` argument is newly added to the functions.

**Review by ammaraskar — APPROVED** [review]
LGTM with suggestions. 

Thanks for the patch Anthony. Confirmed that 3.6 seems to be the first version path objects started working. It would be helpful to introduce this test to ensure that it doesn't accidentally break over time.

File: Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst (hunk 4 above)
Comment (zware) [reply Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst]: +1 on this comment, LGTM otherwise.

File: Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Document and test that ``tempfile`` functions may accept a
+:term:`path-like object` for the ``dir`` argument.  Patch by Anthony Sottile.
+is a :term:`path-like object`.  Patch by Anthony Sottile.
```
Comment (zware) [inline Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst]: The suggestion went awry here; this last line needs to go away.

Comment (asottile) [reply Misc/NEWS.d/next/Documentation/2019-01-21-14-30-59.bpo-35803.yae6Lq.rst]: oops! fixed

**Review by zware — APPROVED** [review]
LGTM

**Comment by zware** [comment]
Thanks for the patch and reviews!
